man/tpm2_getrandom.1.in
.\" Copyright (c) 2016, Intel Corporation
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright notice,
.\" this list of conditions and the following disclaimer.
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright notice,
.\" this list of conditions and the following disclaimer in the documentation
.\" and/or other materials provided with the distribution.
.\"
.\" 3. Neither the name of Intel Corporation nor the names of its contributors
.\" may be used to endorse or promote products derived from this software without
.\" specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
.\" THE POSSIBILITY OF SUCH DAMAGE.
.TH tpm2_getrandom 8 "DECEMBER 2016" Intel "tpm2-tools"
.SH NAME
tpm2_getrandom\ - returns the next SIZE octets from the random number generator.
.SH SYNOPSIS
.B tpm2_getrandom[ COMMON OPTIONS ] [ TCTI OPTIONS ] [ -\-output\fR|\fB ] SIZE
.PP
returns the next SIZE octets from the random number generator.
.SH DESCRIPTION
.B tpm2_getrandom
returns the next SIZE octets from the random number generator.
.SH OPTIONS
.TP
\fB\-o ,\-\-output\fR
specifies the filename of output.
@COMMON_OPTIONS_INCLUDE@
@TCTI_OPTIONS_INCLUDE@
.SH ENVIRONMENT\@TCTI_ENVIRONMENT_INCLUDE@
.SH EXAMPLES
.B tpm2_getrandom
.PP
.nf
.RS
tpm2_getrandom -o random.out 20
tpm2_getrandom 8
.RE
.fi


CHANGELOG
next
* tpm2_create - options 'o' and 'O' changed to 'u' and 'r' respectively. 
* tpm2_pcrlist: support yaml output for parsing.
* tpm2_pcrevent: new tool for hashing and extending pcrs.
* Make tpm2_{createprimary,create,load,pcrlist,hmac} tools to support the --quiet option.
* Support for a --quiet option to suppress messages printed by tools to standard output.
* tpm2_hmac: support for files greater than 1024 bytes, changes in options and arguments.
* tpm2_hash: support for files greater than 1024 bytes, changes in options and arguments.
* Install is now to bin vs sbin. Ensure that sbin tools get removed!
* make dist and distcheck are now working.
* installation into customized locations are now working, see issue #402 for details.
* tpm2_pcrlist: renamed from tpm2_listpcrs.
* tpm2_pcrextend: new tool for extending PCRs.
* tpm2_getmanufec: -E option no longer required, defaults to stdout.
* tpm2_nvlist: Support for friendly nv attributes in output.
* Support for friendly algorithm names for algorithm identifiers.
* tpm2_nvread: The option, -s, or size option is no longer required.
* tpm2_nvread: ouutput is now in xxd -r compatible format.
* tpm2_nvwrite: fixed to write files larger than 1024 in size.
* tpm2_nvread: fixed to read files larger than 1024 in size.
* tpm2_nvdefine supports "nice-names" for nv space attributes.
* Support using PCR Policy directly with tpm2_unseal tool.
* Support PCR policy creation in tpm2_createpolicy
* Support using a policy session as input to tools that may need to satisfy complex policies
  other than password.
* tpm2_unseal: supports output to stdoud.
* tpm2_create: enforce policy based authorization.
* tpm2_createprimary: add ability to create objects with policy based authorization.
* tpm2_nvdefine: add ability to create nv indexes with policy based authorization.
* Support Clang Build.
* tpm2_unseal test uses endorsement hierarchy as platform hierarchy is unavailable on a
  real tpm.
* Numerous cleanups and minor bug fixes.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


src/tpm2_getrandom.cpp
//**********************************************************************;
// Copyright (c) 2015, Intel Corporation
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice,
// this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice,
// this list of conditions and the following disclaimer in the documentation
// and/or other materials provided with the distribution.
//
// 3. Neither the name of Intel Corporation nor the names of its contributors
// may be used to endorse or promote products derived from this software without
// specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
// THE POSSIBILITY OF SUCH DAMAGE.
//**********************************************************************;

#include <stdarg.h>

#include <stdio.h>
#include <stdlib.h>   // Needed for _wtoi
#include <string.h>
#include <limits.h>
#include <getopt.h>
#include <ctype.h>

#include <sapi/tpm20.h>
#include <tcti/tcti_socket.h>
#include "common.h"

int debugLevel = 0;

void showHelp(const char *name)
{
    printf("\n%s  [options]\n"
            "Usage: %s  [-h/--help]\n"
            "   or: %s [-v/--version]\n"
            "   or: %s [-s/--size <bytesRequested>] [-o/--of <outfilename>] \n"
            "   or: %s  [-s/--size <bytesRequested>] [-o/--of <outfilename>]\n"
            "                      [-i/--ip <ipAddress>] [-p/--port <port>] [-d/--dbg <dbgLevel>]\n"
            "\nwhere:\n\n"
            "   -h/--help                       display this help and exit.\n"
            "   -v/--version                    display version information and exit.\n"
            "   -s/--size <bytesRequested>      specifies the size of the bytesRequested.\n"
            "   -o/--of <outfilename>           specifies the filename of output:\n"
            "   -p/--port <port>                specifies the port number (default:%d).\n"
            "   -d/--dbg <dbgLevel>             specifies level of debug messages:\n"
            "                                     0 (high level test results)\n"
            "                                     1 (test app send/receive byte streams)\n"
            "                                     2 (resource manager send/receive byte streams)\n"
            "                                     3 (resource manager tables)\n"
            "\nexample:\n"
            "   %s -s 20 -o random.out \n"
            , name, name, name, name, name, DEFAULT_RESMGR_TPM_PORT, name);
}

int getRandom(const char *outFileName, UINT16 bytesRequested)
{
    TPM_RC rval;
    TPM2B_DIGEST        randomBytes = { { sizeof(TPM2B_DIGEST)-2, } };

    rval = Tss2_Sys_GetRandom(sysContext, NULL, bytesRequested, &randomBytes, NULL);
    if (rval != TSS2_RC_SUCCESS)
    {
        printf("\n......TPM2_GetRandom Error. TPM Error:0x%x......\n", rval);
        return -1;
    }
    printf("\nGetRandom succ...\n");
    printf("byte size: %d\n",randomBytes.t.size);
    for(UINT16 i = 0; i < randomBytes.t.size; i++)
        printf(" 0x%2.2X",randomBytes.t.buffer[i]);
    printf("\n");

    if(saveDataToFile(outFileName, (UINT8 *)randomBytes.t.buffer, randomBytes.t.size))
        return -2;

    return 0;
}

int main(int argc, char* argv[])
{
    UINT16 size=20;
    char hostName[200] = DEFAULT_HOSTNAME;
    int port = DEFAULT_RESMGR_TPM_PORT;

    int returnVal = 0;
    char outFileName[PATH_MAX] = {0};

    setbuf(stdout, NULL);
    setvbuf (stdout, NULL, _IONBF, BUFSIZ);

    const char *optstring = "s:o:p:d:hv";
    int opt = -1;

    static struct option long_Opts[] = {
        { "size",1,NULL,'s' },
        { "of",1, NULL,'o' },
        { "port",1,NULL,'p' },
        { "dbg", 1,NULL,'d' },
        { "help",0,NULL,'h' },
        { "version",0,NULL,'v' },
        { 0,0,0,0 },
    };

    int flagCnt = 0,
        s_flag=0,
        o_flag=0,
        h_flag=0,
        v_flag=0;

    if( argc > (int)(2*sizeof(long_Opts)/sizeof(struct option)) || argc == 1)
    {
        showArgMismatch(argv[0]);
        return -1;
    }

    while((opt = getopt_long(argc,argv,optstring,long_Opts,NULL)) != -1 && returnVal == 0)
    {
        switch (opt) {
        case 'h':
            h_flag = 1;
            //  showHelp();
            break;
        case '?':
            returnVal = -2;
            break;
        case 'v':
            //      showVersion();
            v_flag = 1;
            break;

        case 's':
            if(getSizeUint16(optarg, &size))
            {
                printf(" s must go with the size number.\n");
                returnVal = -3;
                break;
            }
            s_flag = 1;
            break;

        case 'o':
            if (optarg==NULL)
            {
                printf("\n Please don't forget to specified the output file! \n");
                returnVal = -4;
                break;
            }
            safeStrNCpy(outFileName, optarg, sizeof(outFileName));
            o_flag=1;
            break;
        case 'p':
            if( getPort(optarg, &port) )
            {
                printf("Incorrect port number.\n");
                returnVal = -5;
            }
            break;
        case 'd':
            if( getDebugLevel(optarg, &debugLevel) )
            {
                printf("Incorrect debug level.\n");
                returnVal = -6;
            }
            break;

        default:
            showArgMismatch(argv[0]);
            exit(-7);
        }
    };

    if(returnVal != 0)
        return returnVal;

    flagCnt = h_flag + v_flag + o_flag + s_flag;
    if (flagCnt == 1)
    {
        if(h_flag == 1)
            showHelp(argv[0]);
        else if(v_flag == 1)
            showVersion(argv[0]);
        else
        {
            showArgMismatch(argv[0]);
            return -8;
        }
    }
    else if(flagCnt == 2 && h_flag != 1 && v_flag !=1)
    {
        prepareTest(hostName, port, debugLevel);

        returnVal = getRandom(outFileName, size);

        finishTest();

        if(returnVal)
            return -9;
    }
    else
    {
        showArgMismatch(argv[0]);
        return -10;
    }

    return 0;
}
